### PR TITLE
Move irrational constants to its own module

### PR DESCRIPTION
**Pull request by fredrikekre**

### CODE REVIEW

File: base/irrationals.jl [hunk 1]
```diff
@@ -142,94 +142,5 @@ end
...
-# but need to loop over types to prevent ambiguity with generic rules for ^(::Number, x) etc.
-for T in (Irrational, Rational, Integer, Number)
-    ^(::Irrational{:e}, x::T) = exp(x)
-end
-
-log(::Irrational{:e}) = 1 # use 1 to correctly promote expressions like log(x)/log(e)
-log(::Irrational{:e}, x::Number) = log(x)
-
-# align along = for nice Array printing
-function alignment(io::IO, x::Irrational)
```
Comment (giordano) [inline base/irrationals.jl]: Was this method removed on purpose?

Comment (fredrikekre) [inline base/irrationals.jl]: No, should have been moved.

**Comment by Keno** [comment]
This was discussed in triage and the preferred resolution was the following:
- Delete `eu`
- Make `ℯ` (\mscre) a unicode alias for `e` (i.e. rebase #10612)
- Export `π` and `ℯ` from base by default, put the ascii versions as well as the other math constants in its own module (this PR)
Doesn't all have to be done in this PR, but I wanted to note it down somewhere relevant.

**Comment by simonbyrne** [comment]
I think we should probably export `pi` as well.

File: base/linalg/lapack.jl [hunk 2]
```diff
@@ -5299,7 +5299,7 @@ for (bdsdc, elty) in
                 u, &ldu, vt, &ldvt,
                 q, iq, work, iwork, info)
             chklapackerror(info[])
-            d, e, u, vt, q, iq
+            d, e_, u, vt, q, iq
```
Comment (fredrikekre) [inline base/linalg/lapack.jl]: Apparently we returned that mathematical constant `e` from `LAPACK.bdsdc!`. :metal:

**Comment by fredrikekre** [comment]
At time of writing I think this PR now implements the new behaviour. I think the easiest would be to incorporate #10612 in this PR for the deprecation of `eu` etc. What should we name the new module? `Constants`, `MathConsts`, `MathConstants`, `MathematicalConstants` or other suggestions?

**Comment by giordano** [comment]
Note that ISO mandates that mathematical constants are to be typeset with roman (upright) font, thus using  `ℯ` for the Euler constant is in contrast with this rule.  Of course, Julia *doesn't have* to follow ISO on this.

**Comment by simonbyrne** [comment]
`Irrationals`, since they're all irrational?

**Comment by fredrikekre** [comment]
`Irrationals`, since they're all irrational?

In that case we should probably wrap everything regarding irrationals in that module, not just the constants.

**Comment by andyferris** [comment]
I actually like that we export ASCII `e` (and definitely `pi`)

**Comment by KristofferC** [comment]
Exporting a one letter variable is way too error prone for a general purpose language, in my opinion. I have thrown `e` multiple times, I have had bugs in my code due to missing initializing my own \gamma etc. The cost of someone wanting to use `e` as `exp(1)` having to do an import or using the Unicode variable seems very worthwhile if it avoids putting a one letter variable into all code.

**Comment by andyferris** [comment]
Doesn't the exact same argument apply to `I`? (Are there other one letter ASCII exports?)

**Comment by KristofferC** [comment]
It sort of does, although I guess it is more rare to use `I` (capital letter) in general code than `e`.  Anyway, I think the path forward here is already decided and most the arguments for and against have been made, so perhaps not necessary to rehash them.

**Comment by SimonDanisch** [comment]
I have never used e but use `I` quite a bit and definitely run into problems with it being exported ;) Guess it all depends on the naming style of a particular person!

**Comment by andyferris** [comment]
> Anyway, I think the path forward here is already decided and most the arguments for and against have been made, so perhaps not necessary to rehash them.

Is this referring to:

> This was discussed in triage 

?

**Comment by fredrikekre** [comment]
`I` would be the last single char base export. We could make that `𝐈` (`\mbfI`) or something, but thats a separate discussion.

File: base/linalg/lapack.jl (hunk 2 above)
Comment (andreasnoack) [reply base/linalg/lapack.jl]: Just checked and `e_` is just garbage after the call to `xbdsdc` so maybe just delete it (and adjust `svd`)

**Comment by giordano** [comment]
This is a breaking change, because the type of the Euler's number is different.  This can probably be handled in `Compat.jl` by defining the same constant.

**Comment by ararslan** [comment]
I must say I still don't like the Unicode alias for `e`. It's too visually similar to ASCII `e` which just makes it confusing. I think we should either keep exporting `e` or put it in the math constants module with no exported alias.

**Comment by fredrikekre** [comment]
Could use some guidance on how to deprecate stuff here. Are we OK with just make this breaking? I tried some things e.g.
```julia
export e
e = Irrational{:e}()
Float64(::Irrational{:e}) = (depwarn("e is moved etc...", :Float64); Float64(ℯ))
# ...
```
but then you get a conflict when `using Base.MathConstants`:
```
WARNING: using MathConstants.e in module Main conflicts with an existing identifier.
```
Any way to actually deprecate this?

**Comment by ararslan** [comment]
Should be able to use `@deprecate_binding e MathConstants.e`

**Comment by fredrikekre** [comment]
Hmmmm, I was sure I tried that, but seems to work. Does not give a very helpful message though:

```julia
julia> e
WARNING: Base.e is deprecated.
  likely near no file:0
ℯ = 2.7182818284590...
```

**Comment by PallHaraldsson** [comment]
@Keno "Make ℯ (\mscre)" What does "mscre" stand for anyway? I'm never going to remember this unless I know..

Can we have (also) \euler for "Euler's number"?: https://en.wikipedia.org/wiki/E_%28mathematical_constant%29

https://en.wikipedia.org/wiki/Euler%E2%80%93Mascheroni_constant

**Comment by fredrikekre** [comment]
`\mscre`  = `math-script-e`

**Comment by Keno** [comment]
mscre stands for "math script e". We can and should add another unicode alias. `\e` has been suggested, but it's annoying, because there's a bunch of other completions that start with `\e..`

**Comment by fredrikekre** [comment]
Ok, so lets decide on a name for the module?
This is also a problem:
```julia
julia> e
WARNING: Base.e is deprecated.
  likely near no file:0
ℯ = 2.7182818284590...

julia> e
WARNING: Base.e is deprecated.
  likely near no file:0
ℯ = 2.7182818284590...
```
The warning is rather cryptic, and does not suggest the alternative and also prints more than once. This works though;
```julia
julia> using Base.MathConstants

julia> e
ℯ = 2.7182818284590...
```
Any ideas would be much appreciated.

File: NEWS.md [hunk 3]
```diff
@@ -363,6 +363,15 @@ Deprecated or removed
 
   * `diagm(A::BitMatrix)` has been deprecated, use `diagm(vec(A))` instead ([#23373]).
 
+  * `ℯ` (written as `\mscre<TAB>` or `\euler<TAB>`) is the new default for Euler's
```
Comment (StefanKarpinski) [inline NEWS.md]: What does it mean for this to be "the new default"? Default in what sense?

**Review by StefanKarpinski — APPROVED** [review]

File: NEWS.md (hunk 3 above)
Comment (fredrikekre) [reply NEWS.md]: The one being exported, and also `typeof(e)` has changed from `Irrational{:e}` to `Irrational{:ℯ}`.

Comment (StefanKarpinski) [reply NEWS.md]: Ok, might be good to clarify that if you're willing!

Comment (fredrikekre) [reply NEWS.md]: Yes, I think my own wording is kinda weird too.

**Pull request merged**